src/cocotb/_event_loop.py

```diff
@@ -0,0 +1,86 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import logging
+import sys
+from collections import deque
+from typing import Callable, Deque
+
+from cocotb import debug
+from cocotb._bridge import run_bridge_threads
+from cocotb._py_compat import cached_property
+
+if sys.version_info >= (3, 10):
+    from typing import ParamSpec
+
+    P = ParamSpec("P")
+
+
+class ScheduledCallback:
+    __slots__ = ("_func", "_args", "_kwargs", "_cancelled")
+
+    def __init__(
+        self, func: "Callable[P, object]", *args: "P.args", **kwargs: "P.kwargs"
+    ) -> None:
+        self._func = func
+        self._args = args
+        self._kwargs = kwargs
+        self._cancelled: bool = False
+
+    def cancel(self) -> None:
+        self._cancelled = True
+
+
+class EventLoop:
+    def __init__(self) -> None:
+        self._callbacks: Deque[ScheduledCallback] = deque()
+        if debug.debug:
+            self._cycles: int = 0
+
+    @cached_property
+    def log(self) -> logging.Logger:
+        return logging.getLogger("cocotb.event_loop")
+
+    def run(self) -> None:
+        while self._callbacks:
```

Comment:
Shouldn't `_cycles` be set to 0 at the start of `run()`?

Reply:
```suggestion
        self._cycles = 0
        while self._callbacks:
```